ref_liom:
  # Topics
  pointCloudTopic: "/cloud_registered_body"   # Point cloud data
  odomTopic: "/Odometry"                      # IMU pre-preintegration odometry, same frequency as IMU
  imuTopic: "/imu/data"                       # IMU data
  gpsTopic: "odometry/gpsz"                   # GPS odometry topic from navsat, see module_navsat.launch file
  loopTopic: "/loop_closure_detection"
  priorTopic: "/prior_constraints"

  # Frames
  bodyFrame: "body"
  odomFrame: "odom"
  mapFrame: "map"
  
  # submap
  submapFrameNum: 5

  # GPS Settings
  useGpsElevation: false                      # if GPS elevation is bad, set to "false"
  gpsCovThreshold: 2.0                        # m^2, threshold for using GPS data
  poseCovThreshold: 25.0                      # m^2, threshold for using GPS data
  
  # Export settings
  savePCD: false                              # https://github.com/TixiaoShan/LIO-SAM/issues/3
  savePCDDirectory: "/Code/rosbag/ref_liom/"  # in your home folder, starts and ends with "/". Warning: the code deletes "LOAM" folder then recreates it. See "mapOptimization" for implementation

  # Sensor Settings
  sensor: robosense                           # lidar sensor type, 'velodyne' or 'ouster' or 'livox'
  
  # voxel filter paprams
  mappingLeafSize: 0.2                        # default: 0.4 - outdoor, 0.2 - indoor
  
  # CPU Params
  numberOfCores: 8                            # number of cores for mapping optimization
  mappingProcessInterval: 0.1                 # seconds, regulate mapping frequency
  
  # Surrounding map
  surroundingkeyframeAddingDistThreshold: 0.2   # meters, regulate keyframe adding threshold
  surroundingkeyframeAddingAngleThreshold: 0.05  # radians, regulate keyframe adding threshold
  surroundingKeyframeDensity: 2.0               # meters, downsample surrounding keyframe poses   
  surroundingKeyframeSearchRadius: 50.0         # meters, within n meters scan-to-map optimization (when loop closure disabled)

  # Loop closure
  loopClosureEnableFlag: false
  loopClosureFrequency: 0.5                    # Hz, regulate loop closure constraint add frequency
  surroundingKeyframeSize: 50                  # submap size (when loop closure enabled)
  historyKeyframeSearchRadius: 15.0            # meters, key frame that is within n meters from current pose will be considerd for loop closure
  historyKeyframeSearchTimeDiff: 30.0          # seconds, key frame that is n seconds older will be considered for loop closure
  historyKeyframeSearchNum: 25                 # number of hostory key frames will be fused into a submap for loop closure
  historyKeyframeFitnessScore: 0.2             # icp threshold, the smaller the better alignment

  # Visualization
  globalMapVisualizationSearchRadius: 1000.0    # meters, global map visualization radius
  globalMapVisualizationPoseDensity: 10.0       # meters, global map visualization keyframe density
  globalMapVisualizationLeafSize: 0.2           # meters, global map visualization cloud density

ref_provider:
  map_frame: "map"
  body_frame: "body"
  point_cloud_topic: "/ref_liom/submap" 
  odometry_topic: "/ref_liom/aftPgo_odom"
  keyframe_idx_topic: "/ref_liom/submap_key_idx"
  model_poses_topic: "/ref_liom/model_poses"
  search_radius: 50.0
  track_bbox_scale: 1
  segment_num_threshold: 300
  max_opti_num: 30
  max_correspondence_dist: 1
  fitness_threshold: 0.1
  overlap_threshold: 0.25
  voxel_size: 0.3

# Navsat (convert GPS coordinates to Cartesian)
navsat:
  frequency: 50
  wait_for_datum: false
  delay: 0.0
  magnetic_declination_radians: 0
  yaw_offset: 0
  zero_altitude: true
  broadcast_utm_transform: false
  broadcast_utm_transform_as_parent_frame: false
  publish_filtered_gps: false

# EKF for Navsat
ekf_gps:
  publish_tf: false
  map_frame: map
  odom_frame: odom
  base_link_frame: base_link
  world_frame: odom

  frequency: 50
  two_d_mode: false
  sensor_timeout: 0.01
  # -------------------------------------
  # External IMU:
  # -------------------------------------
  imu0: imu_correct
  # make sure the input is aligned with ROS REP105. "imu_correct" is manually transformed by myself. EKF can also transform the data using tf between your imu and base_link
  imu0_config: [false, false, false,
                true,  true,  true,
                false, false, false,
                false, false, true,
                true,  true,  true]
  imu0_differential: false
  imu0_queue_size: 50 
  imu0_remove_gravitational_acceleration: true
  # -------------------------------------
  # Odometry (From Navsat):
  # -------------------------------------
  odom0: odometry/gps
  odom0_config: [true,  true,  true,
                 false, false, false,
                 false, false, false,
                 false, false, false,
                 false, false, false]
  odom0_differential: false
  odom0_queue_size: 10

  #                            x     y     z     r     p     y   x_dot  y_dot  z_dot  r_dot p_dot y_dot x_ddot y_ddot z_ddot
  process_noise_covariance: [  1.0,  0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    1.0,  0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    10.0, 0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0.03, 0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0.03, 0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0.1,  0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0.25,  0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0.25,  0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0.04,  0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0.01, 0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0.01, 0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0.5,  0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0.01, 0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0.01,   0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0.015]
